internal/daemon/api_logs.go

```diff
@@ -0,0 +1,262 @@
+		n, err := strconv.Atoi(nStr)
+		if err != nil {
+			response := statusBadRequest("n must be a valid integer")
+			response.ServeHTTP(w, req)
+			return
+		}
+		numLogs = n
+	}
+
+	// If "services" parameter not specified, fetch logs for all services.
+	if len(services) == 0 {
+		infos, err := r.svcMgr.Services(nil)
+		if err != nil {
+			response := statusInternalError("cannot fetch services: %v", err)
+			response.ServeHTTP(w, req)
+			return
+		}
+		services = make([]string, len(infos))
+		for i, info := range infos {
+			services[i] = info.Name
+		}
+	}
+
+	// Get log iterators by service (and close them when we're done).
+	itsByName, err := r.svcMgr.ServiceLogs(services, numLogs)
+	if err != nil {
+		response := statusInternalError("cannot fetch log iterators: %v", err)
+		response.ServeHTTP(w, req)
+		return
+	}
+	defer func() {
+		for _, it := range itsByName {
+			_ = it.Close()
+		}
+	}()
+
+	// Make sorted list of service names we actually have iterators for.
+	services = services[:0]
+	for name := range itsByName {
+		services = append(services, name)
+	}
+	sort.Strings(services)
+
+	// Create an iterator and log parser for each service.
+	iterators := make([]servicelog.Iterator, len(services))
+	parsers := make([]*servicelog.Parser, len(services))
+	for i, name := range services {
+		iterators[i] = itsByName[name]
+		parsers[i] = servicelog.NewParser(iterators[i], logReaderSize)
+	}
+
+	// Output format is JSON Lines, which doesn't have an official mime type,
+	// but "application/x-ndjson" is what most people seem to use:
+	// https://github.com/wardi/jsonlines/issues/9
+	w.Header().Set("Content-Type", "application/x-ndjson")
+	encoder := json.NewEncoder(w)
+	encoder.SetEscapeHTML(false)
+	requestDone := req.Context().Done()
+	requestStarted := time.Now().UTC()
+
+	// If we'll be following real-time, make a channel and register it with
+	// each of the iterators to be notified when new data comes in.
+	var notification chan bool
+	if follow {
+		notification = make(chan bool, 1)
+		for _, it := range iterators {
+			it.Notify(notification)
+		}
+	}
+
+	// Use a buffered channel as a FIFO for keeping the latest numLogs logs if
+	// request "n" is set (the default).
+	var fifo chan servicelog.Entry
+	if numLogs > 0 {
+		fifo = make(chan servicelog.Entry, numLogs)
+	}
+
+	// Slice of next entries for each service
+	nexts := make([]servicelog.Entry, len(services))
+
+	// Changes to true once we start following
+	following := false
+
+	// Main loop: output earliest log per iteration. Stop when no more logs,
+	// or when request is cancelled if in follow mode.
+	for {
+		// Try to fetch next log from each service (parser/iterator combo).
+		for i, parser := range parsers {
+			if !nexts[i].Time.IsZero() {
+				continue
+			}
+			if parser.Next() {
+				nexts[i] = parser.Entry()
+			} else if parser.Err() != nil {
+				log.Printf("error parsing logs: %v", parser.Err())
+				return
+			} else if iterators[i].Next(nil) {
+				parser.Reset()
```

Comment:
At this point we may have an actual entry inside the parser which ideally should be considered for this round since this was just a buffer size behavior and not a timing one, so the time for the upcoming entry might still be earlier than the others. Just adding something like this here should solve it:
```
if parser.Next() {
    nexts[i] = parser.Entry()
}
```
No need to check errors even. We can pick that up on the next round.

Btw, there's a `Reset()` above which I don't quite understand. Why do we need to explicitly reset the parser, instead of just calling `Next` once more? Isn't it just picking up the next item out of the iterator?

Reply:
Good catch. I've added this and a test for the case in `api_logs_test.go`.

`Parser.Reset` was needed to reset the stored error, which was EOF from the previous call to `Next`. However, this is not really an error case, so I refactored slightly to avoid storing EOF in p.err. This also avoids the need for `Reset`, and we can just call `Next` again as you suggested -- thanks.

Reply:
Small update: I did add the test in `api_logs_test.go` but now with this code being refactored into a background goroutine and a main loop it's hard to test this case. For now I've just removed the test (I know it's working from the test I did add). Not 100% coverage I know, but I think it's a reasonable trade-off.